### PR TITLE
make sumaform terraform v11 compatible

### LINKED ISSUE #285: Support terraform v0.11
The [repository](https://download.opensuse.org/repositories/systemsmanagement:/sumaform/SLE_12_SP3/) mentioned in [readme](https://github.com/moio/sumaform/blob/master/README.md) to download terraform from does not have terraform rpm anymore and building is disabled. I went to use [Virtualization:containers](http://download.opensuse.org/repositories/Virtualization:/containers/SLE_12_SP3/) instead, but there is terraform v0.11 already.

Terraform v0.11 does not work with current configuration examples, primarily because of `version` being reserved word now. ( See https://github.com/hashicorp/terraform/issues/16859 ).
Are there any plans to support new version?

### PR DESCRIPTION
**Pull request by MalloZup**
This pr is not taking care of the update to v11 , it is a prerequisite step that we need to implement.

fix #285 

# Left to do:

# gitlab
https://gitlab.suse.de/galaxy/sumaform-test-runner/merge_requests/41

# backup centos7 or sles img

### communication:

- [ ] send an-email to galaxy-devel about the change of API (version != product_version)

### CODE REVIEW

**Comment by MalloZup** [comment]
ok i think the variable change is done.
```bash
/bin/sumaform$ grep -R "\<version\>"
modules/aws/host/variables.tf:  description = "Main product version (eg. 3.2-released, 3.2-nightly, 3.1-nightly, head)"
modules/libvirt/suse_manager/variables.tf:  description = "Leave default for automatic selection or specify an OS supported by the specified product version"
modules/libvirt/suse_manager_proxy/variables.tf:  description = "Leave default for automatic selection or specify an OS supported by the specified product version"
salt/suse_manager_server/testsuite/empty.xml:<?xml version="1.0"?>
salt/grafana/dashboards/suse_manager.json:  "version": 2
salt/minionswarm/minionswarm.py:        help='Each Minion claims a different version grain')
salt/controller/init.sls:# HACK: currently we need this exact version
salt/controller/init.sls:  - version: 2.0.0
salt/controller/init.sls:  - version: 64.0.3282.167
salt/controller/init.sls:  - version: 64.0.3282.167
TROUBLESHOOTING.md:A: please update your terraform version to >= 0.8.0
TROUBLESHOOTING.md:The error means that there is no such version as "3-stable" for the desired resource. Specifically, all SUSE Manager 3.0 versions have been renamed to get a `3.0-` prefix instead of the old `3-` one to avoid confusion with 3.1, moreover `stable` has been replaced by `released` to take into account beta versions.
.travis.yml:before_script: terraform --version
packer/http/ks_sles-es7.cfg:-perl-version
packer/http/ks_centos7.cfg:-perl-version
packer/LICENSE:      the original version of the Work and any modifications or additions
DESIGN.md: - be immutable (not need to be upgraded). Ideally, they should be based on the initial release repo of the target OS (ie. without any updates). Upgrades will be taken care of by Salt states when needed, at the needed version
help/sumaform-logo-color.svg:<?xml version="1.0" encoding="UTF-8"?>
help/sumaform-logo-color.svg:<svg xmlns="http://www.w3.org/2000/svg" xml:space="preserve" width="88.7848mm" height="20.6236mm" version="1.1" style="shape-rendering:geometricPrecision; text-rendering:geometricPrecision; image-rendering:optimizeQuality; fill-rule:evenodd; clip-rule:evenodd"
help/sumaform-logo-black.svg:<?xml version="1.0" encoding="UTF-8"?>
help/sumaform-logo-black.svg:<svg xmlns="http://www.w3.org/2000/svg" xml:space="preserve" width="89.5412mm" height="20.6235mm" version="1.1" style="shape-rendering:geometricPrecision; text-rendering:geometricPrecision; image-rendering:optimizeQuality; fill-rule:evenodd; clip-rule:evenodd"
help/sumaform-icon-black.svg:<?xml version="1.0" encoding="UTF-8"?>
help/sumaform-icon-black.svg:<svg xmlns="http://www.w3.org/2000/svg" xml:space="preserve" width="33.9594mm" height="33.9592mm" version="1.1" style="shape-rendering:geometricPrecision; text-rendering:geometricPrecision; image-rendering:optimizeQuality; fill-rule:evenodd; clip-rule:evenodd"
help/sumaform-icon-color.svg:<?xml version="1.0" encoding="UTF-8"?>
help/sumaform-icon-color.svg:<svg xmlns="http://www.w3.org/2000/svg" xml:space="preserve" width="33.9594mm" height="33.9594mm" version="1.1" style="shape-rendering:geometricPrecision; text-rendering:geometricPrecision; image-rendering:optimizeQuality; fill-rule:evenodd; clip-rule:evenodd"
README_ADVANCED.md:## Changing product version
README_ADVANCED.md:Some modules have a ``product_version`` variable that determines the software product version. Specifically:
README_ADVANCED.md: * in `suse_manager`, `suse_manager_proxy` etc. `product_version` determines the SUSE Manager product version,
README_ADVANCED.md: * in `minion`, `client`, etc. `product_version` determines the SUSE Manager Tools version.
README_ADVANCED.md:Note: the version of Salt on minions is determined by this value, as Salt is obtained from SUSE Manager Tools repos.
README_ADVANCED.md:Note: on clients and minions only, the version number can be omitted to take the default for the distribution, eg. `released` and `nightly` are legal values.
```

**Comment by MalloZup** [comment]
@moio  i'm giving a try with the testsuite.tf. Sofar everything is ok :sun_with_face: 

Note for review: i have splitted the commits in topics so it can be easy for you to go over each commit and review it

File: README_ADVANCED.md [hunk 1]
```diff
@@ -1,10 +1,10 @@
 # Advanced `main.tf` configurations
 
-## Changing product versions
+## Changing product version
```
Comment (moio) [inline README_ADVANCED.md]: versions was OK here

File: README_ADVANCED.md [hunk 2]
```diff
@@ -191,11 +191,11 @@ module "min" {
 }
 ```
 
-## Change the base OS for supported SUSE Manager versions
+## Change the base OS for supported SUSE Manager product_versions
```
Comment (moio) [inline README_ADVANCED.md]: versions was OK here

File: README_ADVANCED.md [hunk 3]
```diff
@@ -191,11 +191,11 @@ module "min" {
...
 
-You can specifiy a base OS for `suse_manager` modules by specifying an `image` variable. There is a default selection if nothing is specified. Please note that not all SUSE Manager versions support all OSs, refer to official documentation to select a compatible OS. In particular, `opensuse423` can presently only be used for the `head` version.
+You can specifiy a base OS for `suse_manager` modules by specifying an `image` variable. There is a default selection if nothing is specified. Please note that not all SUSE Manager product_versions support all OSs, refer to official documentation to select a compatible OS. In particular, `opensuse423` can presently only be used for the `head` product_version.
```
Comment (moio) [inline README_ADVANCED.md]: Please double check all documentation or other non-code places. `product_versions` should be only changed in code, examples, and explicit references to code.

File: modules/openstack/minion/variables.tf [hunk 4]
```diff
@@ -8,8 +8,8 @@ variable "name" {
   type = "string"
 }
 
-variable "version" {
-  description = "A valid SUSE Manager version (eg. 3.0-nightly, head) see README_ADVANCED.md"
+variable "product_version" {
+  description = "A valid SUSE Manager product_version (eg. 3.0-nightly, head) see README_ADVANCED.md"
```
Comment (moio) [inline modules/openstack/minion/variables.tf]: version was correct here

File: README_ADVANCED.md (hunk 3 above)
Comment (MalloZup) [reply README_ADVANCED.md]: yop

File: modules/openstack/minion/variables.tf (hunk 4 above)
Comment (MalloZup) [reply modules/openstack/minion/variables.tf]: @moio  ok i can check the stuff, is really not easy to check all the stuff manually :face_with_head_bandage:

**Comment by MalloZup** [comment]
@juliogonzalez feel free to review. i will attach more details.  soon.

I just added you and MC for backup the eyes of @moio on this, since is a tidy task :+1:

**Comment by juliogonzalez** [comment]
@MalloZup: Ok, I will review tomorrow with a fresh mind and eyes.

File: README_ADVANCED.md [hunk 5]
```diff
@@ -2,9 +2,9 @@
 
 ## Changing product versions
 
-Some modules have a ``version`` variable that determines the software version. Specifically:
- * in `suse_manager`, `suse_manager_proxy` etc. `version` determines the SUSE Manager product version,
- * in `minion`, `client`, etc. `version` determines the SUSE Manager Tools version.
+Some modules have a ``product_version`` variable that determines the software product version. Specifically:
```
Comment (moio) [inline README_ADVANCED.md]: Double backticks here

File: modules/aws/evil_minions/main.tf [hunk 6]
```diff
@@ -47,7 +47,7 @@ hostname: ${replace("${element(aws_instance.instance.*.private_dns, count.index)
 domain: ${var.region}.compute.internal
 use_avahi: False
 mirror: ${var.mirror_private_name}
-version: 3.1-released
+product_version: 3.1-released
```
Comment (juliogonzalez) [inline modules/aws/evil_minions/main.tf]: I am just curious: Why is the version hardcoded here?

Comment (MalloZup) [reply modules/aws/evil_minions/main.tf]: @juliogonzalez afaik because the evil_minion module works on 3.1 but was not tested in head, not sure :thinking:

**Comment by MalloZup** [comment]
@moio  about the terraform pkg, i did a SR. however the pkgs in virtualization are not always building, because the upstream tests are failing. What is your take , should we disable them ? to me seem there is a race condition on the tests that something works sometimes not

File: main.tf.openstack-testsuite.example [hunk 7]
```diff
@@ -1,5 +1,5 @@
 provider "openstack" {
-  version = "~> 1.2"
+  product_version = "~> 1.2"
```
Comment (moio) [inline main.tf.openstack-testsuite.example]: This is wrong, as `version` is an internal variable of the provider in this case.

File: main.tf.openstack.example [hunk 8]
```diff
@@ -1,5 +1,5 @@
 provider "openstack" {
-  version = "~> 1.2"
+  product_version = "~> 1.2"
```
Comment (moio) [inline main.tf.openstack.example]: Same here

**Comment by MalloZup** [comment]
Is safe to accept the pkg, i have tracked the need_research task https://github.com/dmacvicar/terraform-provider-libvirt/issues/342

**Comment by MalloZup** [comment]
# what is left ?
everything is done.  All the needed PRs and SR are in the description.

- [ ] tests sumaform with latest pkgs terraform -v11 and latest libvirt pkg.
- [ ] communicate this via e-mail

Since i will go to holiday next week, @moio you can tests. ( I don't think we will have surprises because docker team already use the 11 terraform and the plugin is more stable now then before :heart: )
:sun_with_face: :beach_umbrella:

**Comment by MalloZup** [comment]
@moio feel free to take my branch and continue the work from there :vulcan_salute:

**Comment by Bischoff** [comment]
@moio I can do the tests if you wish. I'm not sure I can work on MalloZup's branch though.

@moio I can see a red cross with "moio requested changes", but as far as I can see these changes are implemented. Maybe you could mark your review as ok now?

**Comment by Bischoff** [comment]
I can't find Leap 15.0 in http://download.opensuse.org/repositories/home:/SilvioMoioli:/tools/ . I think we need to add this first. Is that thought correct @moio ?

**Comment by Bischoff** [comment]
Bad news.

With these versions:
```
libvirt                    | 4.0.0-lp150.7.3.1 | openSUSE-15.0-Update                                        
terraform                  | 0.11.7-lp150.1.3  | Virtualization:containers (openSUSE_Leap_15.0)              
terraform-provider-libvirt | 0.4.1-lp150.6.1   | Releases for terraform-provider-libvirt (openSUSE_Leap_15.0)
```

from the following repositories
```
openSUSE-15.0-Update                                         | http://download.opensuse.org/update/leap/15.0/oss/                                                               
Virtualization:containers (openSUSE_Leap_15.0)               | http://download.opensuse.org/repositories/Virtualization:/containers/openSUSE_Leap_15.0/                         
Releases for terraform-provider-libvirt (openSUSE_Leap_15.0) | http://download.opensuse.org/repositories/home:/dmacvicar:/terraform-provider-libvirt:/stable/openSUSE_Leap_15.0/
```

terraform crashes:
```
panic: runtime error: invalid memory address or nil pointer dereference
2018-07-30T18:17:30.684+0200 [DEBUG] plugin.terraform-provider-libvirt: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5634b4e71df3]
2018-07-30T18:17:30.684+0200 [DEBUG] plugin.terraform-provider-libvirt: 
2018-07-30T18:17:30.684+0200 [DEBUG] plugin.terraform-provider-libvirt: goroutine 203 [running]:
2018-07-30T18:17:30.684+0200 [DEBUG] plugin.terraform-provider-libvirt: github.com/dmacvicar/terraform-provider-libvirt/libvirt.resourceLibvirtDomainRead(0xc4203660e0, 0x5634b54b4260, 0xc420453170, 0x0, 0x0)
2018-07-30T18:17:30.684+0200 [DEBUG] plugin.terraform-provider-libvirt:         /home/abuild/rpmbuild/BUILD/go/src/github.com/dmacvicar/terraform-provider-libvirt/libvirt/resource_libvirt_domain.go:837 +0x2963
2018-07-30T18:17:30.684+0200 [DEBUG] plugin.terraform-provider-libvirt: github.com/dmacvicar/terraform-provider-libvirt/libvirt.resourceLibvirtDomainCreate(0xc4203660e0, 0x5634b54b4260, 0xc420453170, 0x0, 0x0)
2018-07-30T18:17:30.684+0200 [DEBUG] plugin.terraform-provider-libvirt:         /home/abuild/rpmbuild/BUILD/go/src/github.com/dmacvicar/terraform-provider-libvirt/libvirt/resource_libvirt_domain.go:503 +0x11d9
```

My personal guess is that the provider does not work with a recent terraform.

**Comment by moio** [comment]
That to me looks like a terraform-provider-libvirt bug. Line is the following:

https://github.com/dmacvicar/terraform-provider-libvirt/blob/master/libvirt/resource_libvirt_domain.go#L837

A segmentation violation could arise if any of `networkInterdaceDef`, `Source` or `Network` is `nil` in the `networkInterfaceDef.Source.Network.Network` expression. Question now is: is it correct for that value to be missing in some cases, or do we have a problem in our specific `main.tf`?

Please let's investigate that further in chat today and turn it into a `terraform-provider-libvirt` bug report if necessary. Thanks!

**Comment by Bischoff** [comment]
Problem seems to come from mutualy exclusive "bridge" and "network" interface settings.

Duncan removed this assignement in this PR: dmacvicar/terraform-provider-libvirt#348 and I am currently testing it.

**Comment by Bischoff** [comment]
Now it works with latest version of dmacvicar/terraform-provider-libvirt#348 .

Unfortunately, this means that the Service Request has to be redone after the PR is merged.

**Comment by Bischoff** [comment]
Service requests redone, but now this branch is out of sync with latest commits to sumaform.

**Comment by MalloZup** [comment]
@moio on this codebase, namely SUmaform why your review is RED? can you approve the PR now?

i think this can be potentially merged.

**Comment by moio** [comment]
The only remaining issues are the `destroy` and `volume not found` issues on terraform-provider-libvirt. I will not proceed with merging this PR until they are addressed, or that will disrupt users.

**Comment by MalloZup** [comment]
after the issue we fixed, i cannot find any additional issue.

We can move to TF11 on Monday to me @juliogonzalez cc

**Comment by juliogonzalez** [comment]
@MalloZup that should be possible, but please don't start doing it on Monday until I can confirm.

**Comment by MalloZup** [comment]
@juliogonzalez yop i will cordinate nopb

**Comment by MalloZup** [comment]
@juliogonzalez @moio  @Bischoff so we have a pb with the bridge mode ( which is used in prod)

the https://github.com/dmacvicar/terraform-provider-libvirt/commit/512713f4fd904f8d3e8669b4a9a09130cbe9e3be

is removing the qemu-agent by default which cause for bridged networing do don't be able to retrieve IP adresss of the domain. ANd we have terraform failure so nothing works :sun_with_face: 

Good things is that i found it out before we upgrade

Here we have 2 solution:

- [ ] using env variable for using qemu-agent ( the commit let the user define the variable)

- [ ] or fix the plugin to have qemu-agent enabled when bridge is active ( this can be done on provider in longterm is what should be done afaik) but atm i dunno

**Comment by MalloZup** [comment]
@moio to me we should revert the qemu-agent beeing enabled by default, the risk is to high that we have other networking configuration that will again break.

Follow-up discussion is here:
https://github.com/dmacvicar/terraform-provider-libvirt/pull/373

**Comment by MalloZup** [comment]
https://github.com/dmacvicar/terraform-provider-libvirt/pull/377 
i tested this with sumaform and update also sumaform for using the latest change (https://github.com/moio/sumaform/pull/387/commits/740c5ab7ecd30910f38e9e7ad9e135b2299acacb)

everything was working

**Pull request merged**